Unify LLK HW Configs

ncvetkovicTT opened the pull request:
### Ticket
#33824 

### Problem description
We have many different llk-api HW configs for different threads doing the same thing.

What's changed
This unification effort consolidates multiple specialized HW configuration functions into three main functions:

1. **`llk_math_hw_configure`**: Configures math unit data formats and INT8 math mode
2. **`llk_pack_hw_configure`**: Configures packer with comprehensive tile and format settings  
3. **`llk_unpack_hw_configure`**: Configures both unpackers with format and dimension settings

**Key Benefits:**
- **Reduced LLK API complexity**: Fewer HW configuration calls required
- **Cleaner interface**: Consolidated parameters with sensible defaults
- **Full backward compatibility**: All existing functionality preserved

**Specialized functions** handle specific features:
- **`llk_unpack_configure_stoch_rnd`**: Stochastic rounding configuration (both architectures)
- **`llk_math_reconfig_remap`**: Address remapping for untilize operations (BH only)

This provides a cleaner, more maintainable LLK API while preserving full functionality across both WH and BH architectures.

Apart from necessary API headers, the PR touches the places of the old specialized HW configure calls. All functionality has been preserved.

See tt-llk PR for more details: https://github.com/tenstorrent/tt-llk/pull/926

### Checklist
- [x] [All post commit](https://github.com/tenstorrent/tt-metal/actions/workflows/all-post-commit-workflows.yaml) CI passes - https://github.com/tenstorrent/tt-metal/actions/runs/20414887942
- [x] [Blackhole Post commit](https://github.com/tenstorrent/tt-metal/actions/workflows/blackhole-post-commit.yaml) CI with demo tests passes (if applicable) - https://github.com/tenstorrent/tt-metal/actions/runs/20407507689
- [x] [Model regression](https://github.com/tenstorrent/tt-metal/actions/workflows/perf-models.yaml) CI passes (if applicable) - https://github.com/tenstorrent/tt-metal/actions/runs/20415564396/ 
- [x] [Device performance regression](https://github.com/tenstorrent/tt-metal/actions/workflows/perf-device-models.yaml) CI passes (if applicable) - https://github.com/tenstorrent/tt-metal/actions/runs/20414891270 
- [ ] (For models and ops writers) [Single-card demo tests](https://github.com/tenstorrent/tt-metal/actions/workflows/single-card-demo-tests.yaml) CI passes (if applicable) See [recommended dev flow](https://github.com/tenstorrent/tt-metal/blob/main/models/docs/MODEL_ADD.md#a-recommended-dev-flow-on-github-for-adding-new-models).
- [ ] [Galaxy quick](https://github.com/tenstorrent/tt-metal/actions/workflows/galaxy-quick.yaml) CI passes (if applicable)
- [ ] [Galaxy demo tests, for Llama](https://github.com/tenstorrent/tt-metal/actions/workflows/galaxy-demo-tests.yaml) CI passes, if applicable, because of current Llama work
- [ ] (For runtime and ops writers) [T3000 unit tests](https://github.com/tenstorrent/tt-metal/actions/workflows/t3000-unit-tests.yaml) CI passes (if applicable, since this is run on push to main)
- [ ] (For models and ops writers) [T3000 demo tests](https://github.com/tenstorrent/tt-metal/actions/workflows/t3000-demo-tests.yaml) CI passes (if applicable, since this is required for release)
- [ ] New/Existing tests provide coverage for changes

ncvetkovicTT committed:
#33824: Remove unused parameters from llk-lib:
- DstTileLayout
- ParallelPackerMode
- PackSelMask

ncvetkovicTT committed:
#33824: Remove DstTileFaceLayout:
- Potential bug in _llk_init_packer_dest_offset_registers_

ncvetkovicTT committed:
#33824: Remove write_tile_header...
and adapt the calls to llk-lib functions

ncvetkovicTT committed:
#33824: Fix _llk_init_packer_dest_offset_registers_ calls

ncvetkovicTT committed:
#33824: Refactor _llk_math_hw_configure_:
- Remove skip_inputs in BH implementation
- Remove row_pool in WH implementation
- Make new funtion for DEST remapping _llk_math_reconfig_remap_

ncvetkovicTT committed:
#33824: Remove unused _llk_pack_reduce_hw_configure_...
and corresponding higher-level API functions

ncvetkovicTT committed:
#33824: PACK & UNPACK HW Config (Matmul and AB)
- Remove APIs for _llk_unpack_AB_matmul_hw_configure_
- Rename APIs for _llk_unpack_AB_hw_configure_ to _llk_unpack_hw_configure_
- Remove APIs for _llk_pack_reduce_hw_configure_

#0: Minor fix in tilize_uninit

#0: Fix params in llk_unpack_hw_configure

#0: Fix matmul and pack_dest_init API calls

Fix UMD submodule to match main

ncvetkovicTT committed:
#33824: Update matmul partial face and int8 math WH

ncvetkovicTT committed:
#33824: Remove unpack_A_hw_configure

#0: Fix matmul partial face

ncvetkovicTT committed:
#33824: Add disable_src_zero_flag to unpack_A_init

#0: Fix incorrect hw configure for :
transpose_wh_init_short and test_untilize_calculate_untilize_L1 (LLK)

#0: Fix BH disable_src_zero_flag in unpack_A_init

ncvetkovicTT committed:
#33824: Remove llk_unpack_reduce_hw_configure variants

ncvetkovicTT committed:
#33824: Remove llk_unpack_tilize_hw_configure variants

ncvetkovicTT committed:
#33824: Remove llk_unpack_untilize_hw_configure variants

ncvetkovicTT committed:
#33824: Remove llk_unpack_fast_tilize_hw_configure variants

ncvetkovicTT committed:
#33824: Remove llk_unpack_tilize:
- llk_unpack_tilizeA_B_hw_configure
- llk_unpack_tilizeA_B_hw_configure_disaggregated

#33824: Remove TODOs

#0: Fix matmul init comment

#0: Fix partial face check in matmul

#0: Fix a comment in llk_unpack_hw_configure

pavlejosipovic reviewed (commented):
pls post passing pipelines for perf

nvelickovicTT commented on tt_metal/hw/ckernels/wormhole_b0/metal/llk_api/llk_pack_api.h:
```diff
@@ -23,6 +23,7 @@
  * LLK PACK
  *************************************************************************/
 
+// TODO NC: can be removed altogether:
```
Is this planned to stay beyond this PR? If so, please add an issue to it. Otherwise, just a reminder to remove before merging.

nvelickovicTT commented on tt_metal/include/compute_kernel_api/bcast.h:
```diff
@@ -68,7 +69,7 @@ void reconfigure_unary_bcast(uint32_t old_icb, uint32_t new_icb, uint32_t old_oc
     if (unpacker_src_format_change || unpacker_dst_format_change) {
         // Will configure A & B in similar way
-        UNPACK((llk_unpack_A_hw_configure_disaggregated<DST_ACCUM_MODE>(new_icb)));
+        UNPACK((llk_unpack_hw_configure<DST_ACCUM_MODE>(new_icb, new_icb)));
```
Passing the same CB for both A and B implies we are programming both unpackers with the same CB metadata. Which behind the scenes we are, but to the users it might look contradictory.

Perhaps it would be clearer if we have an overload of `llk_unpack_hw_configure` that only accepts 1 CB as input?
For example:

```
template <bool is_fp32_dest_acc_en>
inline void llk_unpack_hw_configure(const std::uint32_t unpA_operand)
{
   llk_unpack_hw_configure<is_fp32_dest_acc_en>(unpA_operand, unpA_operand);
}
```

nvelickovicTT commented on tt_metal/include/compute_kernel_api/bcast.h:
```diff
@@ -203,20 +204,20 @@ void init_bcast(uint32_t icb0, uint32_t icb1, uint32_t ocb) {
     // clone of general init for AB TODO(AP): commonize
     // UNPACK(( llk_unpack_AB_init<BroadcastType::NONE>() ));
-    // UNPACK(( llk_unpack_AB_hw_configure_disaggregated<DST_ACCUM_MODE>(icb0, icb1) ));
+    // UNPACK(( llk_unpack_hw_configure<DST_ACCUM_MODE>(icb0, icb1) ));
```
Remove dead code.

ncvetkovicTT committed:
#0: UnpackToDest enum fix and partial_tile in matmul

ncvetkovicTT commented on tt_metal/hw/ckernels/wormhole_b0/metal/llk_api/llk_unpack_AB_matmul_api.h:
```diff
@@ -89,7 +53,9 @@ __attribute__((always_inline)) inline void llk_unpack_AB_matmul_init(
         partial_face_a,
-        partial_face_b);
+        partial_face_b,
+        unpA_tile_size,
```
Correct. I will create an issue though, because I can imagine that this API is somewhere used incorrectly, perhaps passing the desired tile size without setting other needed params, resulting in correct outputs for particular cases or data patterns:
https://github.com/tenstorrent/tt-metal/issues/34495

ncvetkovicTT committed:
#0: Rename tile size param for matmul unpack and
add a comment about the future cleanup

ncvetkovicTT committed:
#0: Add comments for future cleanup and remove
dead code and unused functions

ncvetkovicTT committed:
#33824: Add single operand  llk_unpack_hw_configure

ncvetkovicTT committed:
#0: Hard fail, fix llk_unpack_hw_configure single operand case

ncvetkovicTT committed:
#0: Fix matmul_large_block_zm test

ncvetkovicTT committed:
#0: Fix llk_unpack_hw_configure call order

ncvetkovicTT commented:
> pls post passing pipelines for perf

This PR will introduce a slight perf degradation of unet:
https://github.com/tenstorrent/tt-metal/actions/runs/20275850620/job/58224817028

This is because we're trying to make a clear distinction between hardware states that are touched in HW config vs inits.
I removed `llk_unpack_matmul_hw_config` variants, which used to set tile dims for matmul, and moved this setting to `_llk_unpack_AB_matmul_init_`. Since the init is called more often in kernel loops (for unet these kernels are `bmm_large_block_zm_fused_bias_activation` `conv_bmm_tilize`), we're seeing this perf degradation.

Although degrading the perf right now, it enables llk-api contract enforcement which will speed up future LLK feature development and make perf optimizations possible.
FYI @mbezuljTT

ncvetkovicTT commented:
APC: https://github.com/tenstorrent/tt-metal/actions/runs/20308431997
BPC (same as [main base](https://github.com/tenstorrent/tt-metal/actions/runs/20315692568)): https://github.com/tenstorrent/tt-metal/actions/runs/20319137125
Model Perf: https://github.com/tenstorrent/tt-metal/actions/runs/20319144158
Device Perf (failing ResNet50 same as [main](https://github.com/tenstorrent/tt-metal/actions/runs/20320117862), plus slight unet perf degradation which is expected): https://github.com/tenstorrent/tt-metal/actions/runs/20319156224

ncvetkovicTT committed:
#0: Delete setting tile size in reconfigs and
remove unused functions

ncvetkovicTT committed:
#33824: Remove unused tile size from reconfigs

ncvetkovicTT commented:
> > > pls post passing pipelines for perf
> > 
> > 
> > This PR will introduce a slight perf degradation of unet: https://github.com/tenstorrent/tt-metal/actions/runs/20275850620/job/58224817028
> > This is because we're trying to make a clear distinction between hardware states that are touched in HW config vs inits. I removed `llk_unpack_matmul_hw_config` variants, which used to set tile dims for matmul, and moved this setting to `_llk_unpack_AB_matmul_init_`. Since the init is called more often in kernel loops (for unet these kernels are `bmm_large_block_zm_fused_bias_activation` `conv_bmm_tilize`), we're seeing this perf degradation.
> > Although degrading the perf right now, it enables llk-api contract enforcement which will speed up future LLK feature development and make perf optimizations possible. FYI @mbezuljTT
> 
> This is a bit depressing, this means that each matmul and each conv will be a bit slower. Matmul LLK is usually key for perf, and I think regressing it is not OK. Is there some other option for us here (to have a cake and eat it too)

There's another option which I already tried - that's to move tile size setting back to HW config and all reconfig calls. However, we're still below 2890 samples/s. What I saw is that on the newest main this limit went to 2875. This just might make my changes OK, since I somtimes saw 2872, other times 2885 for my code.  Testing it right now.

ncvetkovicTT committed:
#33824: Revert disable_src_zero_flag parameter to
_llk_unpack_hw_configure_

ncvetkovicTT committed:
#0: Fix hw config parameters in LLK tests

ncvetkovicTT committed:
#0: Revert setting tile dims to hw config and
reconfigs

ncvetkovicTT committed:
#0: Fix llk-lib calls for LLK tests and cleanup

ncvetkovicTT committed:
#0: Add missing TODOs with issue numbers

ncvetkovicTT commented:
Waiting for LLK PR to pass PR gate first: https://github.com/tenstorrent/tt-llk/pull/926

ncvetkovicTT committed:
#0: LLK Uplift

fvranicTT commented:
@ncvetkovicTT you will need this workflow too: https://github.com/tenstorrent/tt-metal/actions/runs/20491175868
A bunch of tests have been moved here from APC/BH PC

ncvetkovicTT commented:
> @ncvetkovicTT you will need this workflow too: https://github.com/tenstorrent/tt-metal/actions/runs/20491175868 A bunch of tests have been moved here from APC/BH PC

Thanks for triggering it, I'll keep an eye on it.

ncvetkovicTT commented:
All failures are the same as on main. Merging.
